server/channels/app/desktop_login.go

```diff
@@ -0,0 +1,71 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package app
+
+import (
+	"net/http"
+
+	"github.com/mattermost/mattermost/server/public/model"
+)
+
+func (a *App) SaveClientDesktopToken(token string, createdAt int64) *model.AppError {
+	// Create token in the database
+	err := a.Srv().Store().DesktopTokens().Insert(token, createdAt, nil)
+	if err != nil {
+		return model.NewAppError("SaveClientDesktopToken", "app.desktop_token.create.error", nil, err.Error(), http.StatusInternalServerError)
+	}
+
+	return nil
+}
+
+func (a *App) AuthenticateClientDesktopToken(token string, expiryTime int64, user *model.User) *model.AppError {
+	// Throw an error if the token is expired
+	err := a.Srv().Store().DesktopTokens().SetUserId(token, expiryTime, user.Id)
+	if err != nil {
+		// Delete the token if it is expired
+		a.Srv().Store().DesktopTokens().Delete(token)
+
+		return model.NewAppError("AuthenticateClientDesktopToken", "app.desktop_token.authenticate.invalid_or_expired", nil, err.Error(), http.StatusBadRequest)
+	}
+
+	return nil
+}
+
+func (a *App) GenerateAndSaveServerDesktopToken(clientToken string, expiryTime int64) (*string, *model.AppError) {
```

Comment:
The server_token is stored next to the authN user ID along with a TTL, not next to the desktop token.